cmd/clusterctl/client/tree/discovery.go

```diff
@@ -93,37 +106,110 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 
 		if visible {
 			if machineInfra, err := external.Get(ctx, c, &m.Spec.InfrastructureRef, cluster.Namespace); err == nil {
-				tree.Add(m, machineInfra, ObjectMetaName("MachineInfrastructure"), NoEcho(true))
+				tree.Add(m, machineInfra, ObjectMetaName("MachineInfrastructure"), NoEcho(!options.Echo))
```

Comment:
@fabriziopandini Do you think we'd want to add a placeholder unstructured object in this case as well?

Reply:
might be in a followUp PR, let's keep this one scoped (it is already adding a bunch of flag in the same changeset)